text/3491-remove-implicit-features.md

```diff
@@ -0,0 +1,129 @@
+
+On the next edition, cargo will stop exposing optional dependencies as features
+implicitly, requiring users to add `foo = ["dep:foo"]` if they still want it exposed.
+
+# Motivation
+[motivation]: #motivation
+
+While implicit features offer a low overhead way of defining features,
+- It is easy to overlook using `dep:` when the optional dependency is not intended to be exposed
+  - Making it easy for crate authors to use the wrong syntax and be met with errors ([rust-lang/cargo#10125](https://github.com/rust-lang/cargo/issues/10125))
+  - Potentially breaking people if they are later removed ([rust-lang/cargo#12687)](https://github.com/rust-lang/cargo/pull/12687))
+  - Leading to confusing choices when `cargo add` lists features that look the same (e.g. `cargo add serde` showing `derive` and `serde_derive`)
+  - Leading to confusing errors for callers when they reference the dependency, instead of the feature, and things don't work right
+- Tying feature names to dependency names is a code smell because it ties the API to the implementation
+- It requires people and tools to deal with this special case ([rust-lang/cargo#10543](https://github.com/rust-lang/cargo/issues/10543))
+  - Granted, anything having to deal with old editions will still have to deal with this
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+
+Updated documentation:
+
+## Optional Dependencies
+*(Replaces the [same section in the book](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies))*
+
+Dependencies can be marked "optional", which means they will not be compiled
+by default.
+
+For example, let's say that our 2D image processing library uses
+an external package to handle GIF images. This can be expressed like this:
+
+```toml
+[features]
+gif = ["dep:giffy"]
+
+[dependencies]
+giffy = { version = "0.11.1", optional = true }
+```
+
+This means that this dependency will only be included if the `gif`
+feature is enabled.
+In addition to `cfg(feature = "gif")` syntax, you can reference this as
+`cfg(feature = "giffy")` in the code though users may get confused at error
+messages when used on a public API.
+
+> **Note**: Prior to the 202X edition, features were implicitly created for
+> optional dependencies not referenced via `dep:`.
+
+> **Note**: Another way to optionally include a dependency is to use
+> [platform-specific dependencies]. Instead of using features, these are
+> conditional based on the target platform.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Existing editions
+
+For existing editions, `cargo` will warn when parsing a workspace member's
+package when an optional dependency is not referenced via `dep:` in the
+features ([rust-lang/cargo#9088](https://github.com/rust-lang/cargo/issues/9088)) using the
+planned warning control system ([rust-lang/cargo#12235](https://github.com/rust-lang/cargo/issues/12235)).
```

Comment:
I think it might be helpful to write down a contingency plan if that warning system is not available before the edition needs to be ready. I don't know what the timeline for either the warning system or the edition looks like, so I'm uncertain if the two can line up. But my suggestion would be to add some text along the lines of:

> If the user-controllable warning system isn't implemented in time for the next edition, `cargo fix --edition` will still automatically fix `Cargo.toml` to migrate it to the next edition (just potentially using a less generalized implementation specific for this RFC).

Reply:
Based on our discussion in the team meeting, I re-worked this to focus on the primary end-user affect we are going for (`cargo fix --edition`) in b1c02ae6.